nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -348,73 +351,127 @@ impl DataStore {
+                                        .map(to_db_sled_policy),
+                                ),
+                            )
+                            .filter(
+                                dsl::sled_state
+                                    .eq_any(valid_old_states.iter().copied()),
+                            )
+                            .into_boxed(),
+                        #[cfg(test)]
+                        ValidateTransition::No => query.into_boxed(),
+                    };
+
+                    let query = query
+                        .set((
+                            dsl::sled_policy
+                                .eq(to_db_sled_policy(new_sled_policy)),
+                            dsl::time_modified.eq(Utc::now()),
+                        ))
+                        .check_if_exists::<Sled>(sled_id);
+
+                    let result = query.execute_and_check(&conn).await?;
+
+                    let old_policy = match (check, result.status) {
+                        (ValidateTransition::Yes, UpdateStatus::Updated) => {
+                            result.found.policy()
+                        }
+                        (
+                            ValidateTransition::Yes,
+                            UpdateStatus::NotUpdatedButExists,
+                        ) => {
+                            // Two reasons this can happen:
+                            // 1. An idempotent update: this is treated as a success.
+                            // 2. Invalid state transition: a failure.
+                            //
+                            // To differentiate between the two, check that the new policy
+                            // is the same as the old policy, and that the old state is
+                            // valid.
+                            if result.found.policy() == new_sled_policy
+                                && valid_old_states
+                                    .contains(&result.found.state())
+                            {
+                                result.found.policy()
+                            } else {
+                                return Err(err.bail(
+                                    TransitionError::InvalidTransition {
+                                        current: result.found,
+                                        transition: SledTransition::Policy(
+                                            new_sled_policy,
+                                        ),
+                                    },
+                                ));
+                            }
+                        }
+                        #[cfg(test)]
+                        (ValidateTransition::No, _) => result.found.policy(),
+                    };
+
+                    // When a sled is expunged, the associated disks with that
+                    // sled should also be implicitly set to expunged.
+                    let new_disk_policy = match new_sled_policy {
```

Comment:
Just for my own understanding: This technically doesn't have to be in a transaction, right? It's just that we need someway to perpetually retry to expunge the disks when we expunge the sleds, and this is the most straightforward and robust?

Reply:
Yeah, it could also easily be a CTE (I think those are the only two options to make the UPDATEs of both tables atomic -- doing two separate UPDATEs back-to-back means that the sled could get expunged without  the disk being expunged, if we crashed halfway through).

Reply:
Yeah, that makes sense. I was more thinking that they don't need to be atomic updates at all neccesarrily, since we only soft-delete sleds. We *could* have an RPW that sat there and lazily expunged disks after detecting the sled was expunged. This is all theoretical though, as the implementation that exists looks great to me.